clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2886,24 +2890,34 @@ class SYCLKernelNameTypeVisitor
   void VisitTagDecl(const TagDecl *Tag) {
     bool UnnamedLambdaEnabled =
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
-    if (!Tag->getDeclContext()->isTranslationUnit() &&
-        !isa<NamespaceDecl>(Tag->getDeclContext()) && !UnnamedLambdaEnabled) {
-      const bool KernelNameIsMissing = Tag->getName().empty();
-      if (KernelNameIsMissing) {
+    const DeclContext *DC = Tag->getDeclContext();
+    if (DC && !UnnamedLambdaEnabled) {
+      auto *NS = dyn_cast_or_null<NamespaceDecl>(DC);
+      if (NS && NS->isStdNamespace()) {
         S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << /* kernel name is missing */ 0;
+            << /* kernel name cannot be a type in the std namespace */ 3;
         IsInvalid = true;
       } else {
-        if (Tag->isCompleteDefinition()) {
-          S.Diag(KernelInvocationFuncLoc,
-                 diag::err_sycl_kernel_incorrectly_named)
-              << /* kernel name is not globally-visible */ 1;
-          IsInvalid = true;
-        } else
-          S.Diag(KernelInvocationFuncLoc, diag::warn_sycl_implicit_decl);
-
-        S.Diag(Tag->getSourceRange().getBegin(), diag::note_previous_decl)
-            << Tag->getName();
+        if (!DC->isTranslationUnit() && !isa<NamespaceDecl>(DC)) {
+          const bool KernelNameIsMissing = Tag->getName().empty();
+          if (KernelNameIsMissing) {
+            S.Diag(KernelInvocationFuncLoc,
+                   diag::err_sycl_kernel_incorrectly_named)
+                << /* kernel name is missing */ 0;
+            IsInvalid = true;
+          } else {
+            if (Tag->isCompleteDefinition()) {
+              S.Diag(KernelInvocationFuncLoc,
+                     diag::err_sycl_kernel_incorrectly_named)
+                  << /* kernel name is not globally-visible */ 1;
+              IsInvalid = true;
+            } else
+              S.Diag(KernelInvocationFuncLoc, diag::warn_sycl_implicit_decl);
+
+            S.Diag(Tag->getSourceRange().getBegin(), diag::note_previous_decl)
+                << Tag->getName();
+          }
+        }
       }
     }
```

Comment:
In addition to comment from @elizabethandrews , I suggest to use more early exits in this code (https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code).